### ISSUE TITLE
Black Image After Training and No Training Loss / Moving Tensors From Meta Device Error

### ISSUE THREAD

**Issue by corbin-hayden13** [kijai/ComfyUI-FluxTrainer#36]
I get the following message after generating consistently when using my trained lora: 'D:\ComfyUI\nodes.py:1498: RuntimeWarning: invalid value encountered in cast: img = Image.fromarray(np.clip(i, 0, 255).astype(np.uint8))'

When training I had no loss as my graph images were blank.  Below is my init flux lora training node:
![image](https://github.com/user-attachments/assets/91861bc3-1210-466b-ac91-83233eb42363)

I trained at 512x512 resolution with caption files on 12 gb vram local.

As a separate and/or related issue, there were three separate instances where I got the error `Cannot copy out of meta tensor; no data! Please use torch.nn.Module.to_empty() instead of torch.nn.Module.to() when moving module from meta to a different device.`.  Pasted below is an example of the code where this happened and how I patched it myself.  I have no idea what I'm doing, just tried to get the errors to go away.

```python
    def prepare_split_model(self, model, weight_dtype, accelerator, args):
        from accelerate import init_empty_weights

        logger.info("prepare split model")
        with init_empty_weights():
            flux_upper = flux_models.FluxUpper(model.params)
            flux_lower = flux_models.FluxLower(model.params)
        sd = model.state_dict()

        # lower (trainable)
        logger.info("load state dict for lower")
        flux_lower.load_state_dict(sd, strict=False, assign=True)
        flux_lower.to(dtype=weight_dtype)

        # upper (frozen)
        logger.info("load state dict for upper")
        flux_upper.load_state_dict(sd, strict=False, assign=True)

        logger.info("prepare upper model")
        target_dtype = torch.float8_e4m3fn if args.fp8_base else weight_dtype
        # I added this flux_upper.to_empty() line because I got errors about moving tensors from the meta device to a different device.
        flux_upper.to_empty(device=accelerator.device)
        flux_upper.to(accelerator.device, dtype=target_dtype)
```

Any help on either issue would be greatly appreciated!

**Comment by corbin-hayden13** [comment]
Here is the full error report for the Meta Tensor Error:


# ComfyUI Error Report
## Error Details
- **Node Type:** InitFluxLoRATraining
- **Exception Type:** NotImplementedError
- **Exception Message:** Cannot copy out of meta tensor; no data! Please use torch.nn.Module.to_empty() instead of torch.nn.Module.to() when moving module from meta to a different device.
## Stack Trace
```
  File "D:\ComfyUI\execution.py", line 317, in execute
    output_data, output_ui, has_subgraph = get_output_data(obj, input_data_all, execution_block_cb=execution_block_cb, pre_execute_cb=pre_execute_cb)
                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "D:\ComfyUI\execution.py", line 192, in get_output_data
    return_values = _map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True, execution_block_cb=execution_block_cb, pre_execute_cb=pre_execute_cb)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "D:\ComfyUI\execution.py", line 169, in _map_node_over_list
    process_inputs(input_dict, i)

  File "D:\ComfyUI\execution.py", line 158, in process_inputs
    results.append(getattr(obj, func)(**inputs))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "D:\ComfyUI\custom_nodes\ComfyUI-FluxTrainer\nodes.py", line 430, in init_training
    training_loop = network_trainer.init_train(args)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "D:\ComfyUI\custom_nodes\ComfyUI-FluxTrainer\train_network.py", line 347, in init_train
    model_version, text_encoder, vae, unet = self.load_target_model(args, weight_dtype, accelerator)
                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "D:\ComfyUI\custom_nodes\ComfyUI-FluxTrainer\flux_train_network_comfy.py", line 52, in load_target_model
    model = self.prepare_split_model(model, weight_dtype, accelerator, args)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "D:\ComfyUI\custom_nodes\ComfyUI-FluxTrainer\flux_train_network_comfy.py", line 87, in prepare_split_model
    flux_upper.to(accelerator.device, dtype=target_dtype)

  File "D:\ComfyUI\venv\Lib\site-packages\torch\nn\modules\module.py", line 1173, in to
    return self._apply(convert)
           ^^^^^^^^^^^^^^^^^^^^

  File "D:\ComfyUI\venv\Lib\site-packages\torch\nn\modules\module.py", line 779, in _apply
    module._apply(fn)

  File "D:\ComfyUI\venv\Lib\site-packages\torch\nn\modules\module.py", line 804, in _apply
    param_applied = fn(param)
                    ^^^^^^^^^

  File "D:\ComfyUI\venv\Lib\site-packages\torch\nn\modules\module.py", line 1166, in convert
    raise NotImplementedError(

```
## System Information
- **ComfyUI Version:** v0.2.1-1-gf067ad1
- **Arguments:** main.py
- **OS:** nt
- **Python Version:** 3.11.9 (tags/v3.11.9:de54cf5, Apr  2 2024, 10:12:12) [MSC v.1938 64 bit (AMD64)]
- **Embedded Python:** false
- **PyTorch Version:** 2.3.0+cu121
## Devices

- **Name:** cuda:0 NVIDIA GeForce RTX 4080 Laptop GPU : cudaMallocAsync
  - **Type:** cuda
  - **VRAM Total:** 12878086144
  - **VRAM Free:** 11571036160
  - **Torch VRAM Total:** 0
  - **Torch VRAM Free:** 0

## Logs
```
2024-09-04 11:02:26,948 - root - INFO - Total VRAM 12282 MB, total RAM 48334 MB
2024-09-04 11:02:26,949 - root - INFO - pytorch version: 2.3.0+cu121
2024-09-04 11:02:28,241 - root - INFO - xformers version: 0.0.27.dev792
2024-09-04 11:02:28,242 - root - INFO - Set vram state to: NORMAL_VRAM
2024-09-04 11:02:28,242 - root - INFO - Device: cuda:0 NVIDIA GeForce RTX 4080 Laptop GPU : cudaMallocAsync
2024-09-04 11:02:28,384 - root - INFO - Using xformers cross attention
2024-09-04 11:02:28,790 - root - INFO - [Prompt Server] web root: D:\ComfyUI\web
2024-09-04 11:02:28,793 - root - INFO - Adding extra search path checkpoints ../sd_next\models/Stable-diffusion
2024-09-04 11:02:28,793 - root - INFO - Adding extra search path configs ../sd_next\models/Stable-diffusion
2024-09-04 11:02:28,793 - root - INFO - Adding extra search path vae ../sd_next\models/VAE
2024-09-04 11:02:28,793 - root - INFO - Adding extra search path loras ../sd_next\models/Lora
2024-09-04 11:02:28,793 - root - INFO - Adding extra search path loras ../sd_next\models/LyCORIS
2024-09-04 11:02:28,793 - root - INFO - Adding extra search path upscale_models ../sd_next\models/ESRGAN
2024-09-04 11:02:28,793 - root - INFO - Adding extra search path upscale_models ../sd_next\models/RealESRGAN
2024-09-04 11:02:28,793 - root - INFO - Adding extra search path upscale_models ../sd_next\models/SwinIR
2024-09-04 11:02:28,793 - root - INFO - Adding extra search path embeddings ../sd_next\models/embeddings
2024-09-04 11:02:28,793 - root - INFO - Adding extra search path hypernetworks ../sd_next\models/hypernetworks
2024-09-04 11:02:28,793 - root - INFO - Adding extra search path controlnet ../sd_next\models/ControlNet
2024-09-04 11:02:28,793 - root - INFO - Adding extra search path checkpoints ../ComfyUI\models/checkpoints/
2024-09-04 11:02:28,793 - root - INFO - Adding extra search path ipadapter ../ComfyUI\models/ipadapter/
2024-09-04 11:02:28,793 - root - INFO - Adding extra search path clip ../ComfyUI\models/clip/
2024-09-04 11:02:28,793 - root - INFO - Adding extra search path clip_vision ../ComfyUI\models/clip_vision/
2024-09-04 11:02:28,793 - root - INFO - Adding extra search path configs ../ComfyUI\models/configs/
2024-09-04 11:02:28,795 - root - INFO - Adding extra search path controlnet ../ComfyUI\models/controlnet/
2024-09-04 11:02:28,795 - root - INFO - Adding extra search path embeddings ../ComfyUI\models/embeddings/
2024-09-04 11:02:28,795 - root - INFO - Adding extra search path loras ../ComfyUI\models/loras/
2024-09-04 11:02:28,795 - root - INFO - Adding extra search path upscale_models ../ComfyUI\models/upscale_models/
2024-09-04 11:02:28,795 - root - INFO - Adding extra search path vae ../ComfyUI\models/vae/
2024-09-04 11:02:31,296 - numexpr.utils - INFO - Note: NumExpr detected 32 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 8.
2024-09-04 11:02:31,297 - numexpr.utils - INFO - NumExpr defaulting to 8 threads.
2024-09-04 11:02:32,187 - root - WARNING - Skip D:\ComfyUI\custom_nodes\ComfyUI-Input-Chaining module for custom nodes due to the lack of NODE_CLASS_MAPPINGS.
2024-09-04 11:02:32,243 - root - INFO - Total VRAM 12282 MB, total RAM 48334 MB
2024-09-04 11:02:32,243 - root - INFO - pytorch version: 2.3.0+cu121
2024-09-04 11:02:32,244 - root - INFO - xformers version: 0.0.27.dev792
2024-09-04 11:02:32,244 - root - INFO - Set vram state to: NORMAL_VRAM
2024-09-04 11:02:32,244 - root - INFO - Device: cuda:0 NVIDIA GeForce RTX 4080 Laptop GPU : cudaMallocAsync
2024-09-04 11:02:34,526 - root - WARNING - Skip D:\ComfyUI\custom_nodes\update_all.py module for custom nodes due to the lack of NODE_CLASS_MAPPINGS.
2024-09-04 11:02:34,527 - root - INFO - 
Import times for custom nodes:
2024-09-04 11:02:34,527 - root - INFO -    0.0 seconds (IMPORT FAILED): D:\ComfyUI\custom_nodes\ComfyUI-Input-Chaining
2024-09-04 11:02:34,527 - root - INFO -    0.0 seconds (IMPORT FAILED): D:\ComfyUI\custom_nodes\update_all.py
2024-09-04 11:02:34,527 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\websocket_image_save.py
2024-09-04 11:02:34,527 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\ComfyUI_SimpleMath
2024-09-04 11:02:34,527 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\ComfyUI-Save-Anywhere
2024-09-04 11:02:34,527 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\ComfyUI-Better-Image-Dimensions
2024-09-04 11:02:34,527 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\ComfyUI-DARE-LoRA-Merge
2024-09-04 11:02:34,527 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\TimelineUI_test
2024-09-04 11:02:34,527 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\ComfyUI-CSV-Loader
2024-09-04 11:02:34,527 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\cg-image-picker
2024-09-04 11:02:34,529 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\TimeUi-a-ComfyUi-Timeline-Node
2024-09-04 11:02:34,529 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\ComfyUI_IPAdapter_plus
2024-09-04 11:02:34,529 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\ComfyUI-IPAnimate
2024-09-04 11:02:34,529 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\ComfyUI-OpenPose-Editor
2024-09-04 11:02:34,529 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\comfy-image-saver
2024-09-04 11:02:34,529 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\ComfyUI-Custom-Scripts
2024-09-04 11:02:34,529 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\ComfyUI-GGUF
2024-09-04 11:02:34,529 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\ComfyUI-Frame-Interpolation
2024-09-04 11:02:34,529 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\ComfyUI-Advanced-ControlNet
2024-09-04 11:02:34,529 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\comfyui-prompt-reader-node
2024-09-04 11:02:34,529 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\rgthree-comfy
2024-09-04 11:02:34,529 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\ComfyUI_essentials
2024-09-04 11:02:34,529 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\ComfyUI-AnimateDiff-Evolved
2024-09-04 11:02:34,529 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\ComfyUI_Comfyroll_CustomNodes
2024-09-04 11:02:34,530 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\ComfyUI-KJNodes
2024-09-04 11:02:34,530 - root - INFO -    0.0 seconds: D:\ComfyUI\custom_nodes\ComfyUI-Inspire-Pack
2024-09-04 11:02:34,530 - root - INFO -    0.1 seconds: D:\ComfyUI\custom_nodes\facerestore_cf
2024-09-04 11:02:34,530 - root - INFO -    0.1 seconds: D:\ComfyUI\custom_nodes\ComfyUI_node_Lilly
2024-09-04 11:02:34,530 - root - INFO -    0.2 seconds: D:\ComfyUI\custom_nodes\ComfyUI-Impact-Pack
2024-09-04 11:02:34,530 - root - INFO -    0.3 seconds: D:\ComfyUI\custom_nodes\ComfyUI-VideoHelperSuite
2024-09-04 11:02:34,530 - root - INFO -    0.6 seconds: D:\ComfyUI\custom_nodes\ComfyUI-Manager
2024-09-04 11:02:34,530 - root - INFO -    1.1 seconds: D:\ComfyUI\custom_nodes\comfy_mtb
2024-09-04 11:02:34,530 - root - INFO -    2.7 seconds: D:\ComfyUI\custom_nodes\ComfyUI-FluxTrainer
2024-09-04 11:02:34,530 - root - INFO - 
2024-09-04 11:02:34,540 - root - INFO - Starting server

2024-09-04 11:02:34,540 - root - INFO - To see the GUI go to: http://127.0.0.1:8188
2024-09-04 11:02:55,805 - root - INFO - got prompt
2024-09-04 11:02:55,810 - root - ERROR - Failed to validate prompt for output 130:
2024-09-04 11:02:55,810 - root - ERROR - * FluxTrainEnd 133:
2024-09-04 11:02:55,811 - root - ERROR -   - Required input is missing: network_trainer
2024-09-04 11:02:55,811 - root - ERROR - * AddLabel 78:
2024-09-04 11:02:55,811 - root - ERROR -   - Required input is missing: image
2024-09-04 11:02:55,811 - root - ERROR - * AddLabel 80:
2024-09-04 11:02:55,811 - root - ERROR -   - Required input is missing: image
2024-09-04 11:02:55,811 - root - ERROR - Output will be ignored
2024-09-04 11:02:55,811 - root - ERROR - Failed to validate prompt for output 74:
2024-09-04 11:02:55,811 - root - ERROR - Output will be ignored
```
## Attached Workflow
Please make sure that workflow does not contain any sensitive information such as API keys or passwords.
```
Workflow too large. Please manually upload the workflow from local file system.
```

## Additional Context
(Please add any additional context or steps to reproduce the error here)

**Comment by corbin-hayden13** [comment]
I fixed the meta tensor issue by switching my training Flux model from `flux-dev-fp8` to just `flux-dev` so I think that's what my issue was.
